File: MinimumElement.java
public class MinimumElement {
    //TC is O(logn) as array is divided into half in each iteration
    public static int findMin(int[] a) {
        int low = 0, mid = 0;
        int high = a.length - 1;
        while (low <= high) {
            mid = low + (high - low) / 2;
            if (a[low] <= a[high])
                return a[low];
            else if ((mid == 0 || a[mid - 1] > a[mid]) && (mid == a.length - 1 || a[mid + 1] > a[mid])) //if mid is less than its neighbours
                return a[mid];
            else if (a[low] > a[mid]) { //min element will be present in unsorted array
                high = mid - 1;
            } else {
                low = mid + 1;
            }
        }
        return a[mid];
    }

    public static void main(String args[]) {
        int[] arr = { 4, 5, 6, 7, 0, 1, 2 };
        System.out.printf("Expected 0, got %d", findMin(arr));
    }
}

File: Minimum_element.py
#Time Complexity:- O(log n)
#Space Complexity:- O(1)
class Solution:
    def findMin(self, nums: List[int]) -> int:
        if not nums:  # Handle empty list
            return -1  # Return -1 (or any other suitable value) for no minimum element found

        low = 0  # Set low = 0
        high = len(nums) - 1  # Set high to the last index

        # Check if the array is sorted or not
        if nums[low] <= nums[high]:
            return nums[low]

        # Binary Search
        while low <= high:
            mid = low + (high - low) // 2

            # Check if mid is the minimum element
            if nums[mid] < nums[mid - 1]:
                return nums[mid]

            # Check if mid is greater than high
            if nums[mid] > nums[high]:
                low = mid + 1  # Left half is sorted, move to the right half

            else:
                high = mid - 1  # Right half is sorted, move to the left half

        return -1  # Return -1 for no minimum element found
